e2e/main_test.go

```diff
@@ -0,0 +1,134 @@
+package e2e
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// global client for use by all tests
+var client kubernetes.Interface
+
+// non-configurable for now
+const namespace = "bootkubetesting"
+
+// TestMain handles setup before all tests
+func TestMain(m *testing.M) {
+	var (
+		kubeconfig    = flag.String("kubeconfig", "../hack/quickstart/cluster/auth/kubeconfig", "absolute path to the kubeconfig file")
+		expectedNodes = flag.Int("nodes", 0, "the number of nodes to expect")
+	)
+	flag.Parse()
+
+	if *expectedNodes == 0 {
+		fmt.Println("Please set --nodes flag with the number of nodes in the running cluster")
+		os.Exit(1)
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	// create the clientset
+	client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	if err := ready(client, *expectedNodes); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// createNamespace
+	if _, err := createNamespace(client, namespace); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// run tests
+	exitCode := m.Run()
+
+	if err := deleteNamespace(client, namespace); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	os.Exit(exitCode)
+}
+
+func createNamespace(c kubernetes.Interface, name string) (*v1.Namespace, error) {
+	namespace, err := c.CoreV1().Namespaces().Create(&v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create namespace with name %v %v", name, namespace)
+	}
+	return namespace, nil
+}
+
+func deleteNamespace(c kubernetes.Interface, name string) error {
+	return c.CoreV1().Namespaces().Delete(name, nil)
+}
+
+// Ready blocks until a the cluster is considered available. The current
+// implementation checks that the expected number of nodes are registered.
+func ready(c kubernetes.Interface, expectedNodes int) error {
+	f := func() error {
+		list, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+
+		if len(list.Items) != expectedNodes {
```

Comment:
If we just check for the presence of ready nodes then we won't know if all nodes are online. While waiting for nodes to come up they have no state at all in the api since the apiserver hasn't seen them check in at all yet. 

For example, if I setup terraform-quickstart to run 3 masters and 1 worker and I run `kubectl get nodes` right after `start-cluster.sh` succeeds. I will see that 1 node is ready and no other nodes exist. So I have to know how many nodes to expect. If a node doesn't check-in the cluster isn't in the state you expected and may not have enough nodes for certain tests.